File: src/util/type/ClassName.java
package util.type;

import util.collections.map.ClassMap;

/**
 *
 * @author Martin Polakovic
 */
public class ClassName {

	private final ClassMap<String> namesNoSuper = new ClassMap<>();
	private final ClassMap<String> names = new ClassMap<>();
	private final ClassMap<String> cache = new ClassMap<>();

	/**
	 * Registers name for specified class and through look up all its subclasses that do not
	 * have any name registered.
	 */
	public void add(Class<?> c, String name) {
		names.put(c, name);
	}

	/**
	 * Registers name for specified class. The name/type pair will not partake in lookup.
	 * </p>
	 * This makes sense mostly for overly general classes, such as {@link java.lang.Object} or {@link java.lang.Void}.
	 */
	public void addNoLookup(Class<?> c, String name) {
		namesNoSuper.put(c, name);
	}

	/**
	 * Returns name of the class. If there is no registered name, names of superclasses are looked up
	 * (ignoring the no-look-up entries). If this fails, the name is derived from the class.
	 * <p/>
	 * Name computation is in order:
	 * <ul>
	 * <li> registered name of class
	 * <li> first registered name of superclass in inheritance order
	 * <li> first registered name of interface (no order)
	 * <li> {@link Class#getSimpleName()}
	 * <li> {@link Class#toString()}
	 * </ul>
	 * This is computed lazily (when requested) and cached (every name will be computed only once), with access O(1).
	 *
	 * @return computed name of the class. Never null or empty string.
	 */
	public String get(Class<?> c) {
		cache.computeIfAbsent(c, namesNoSuper::get);
		cache.computeIfAbsent(c, names::getElementOfSuperV);
		return cache.computeIfAbsent(c, ClassName::of);
	}

	public String getOf(Object instance) {
		return get(instance==null ? Void.class : instance.getClass());
	}

	public static String of(Class<?> c) {
		// Anonymous classes don't have human readable names.
		// They "pretend" to be their superclass, handle them as such
		if (c.isAnonymousClass()) return of(c.getSuperclass());

		String n = c.getSimpleName();
		return n.isEmpty() ? c.toString() : n;
	}
}

File: src/util/main/sp/it/util/type/ClassName.kt
package sp.it.util.type

import sp.it.util.collections.map.KClassMap
import sp.it.util.functional.asIs
import kotlin.reflect.KClass
import kotlin.reflect.full.isSuperclassOf
import kotlin.reflect.jvm.jvmName

/**
 * Hierarchical class - className map.
 */
class ClassName {

   private val namesNoSuper = KClassMap<String>()
   private val names = KClassMap<String>()
   private val cache = KClassMap<String>()

   /**
    * Initialization block.
    * All write operations can only be done within this block.
    *
    * This class is not thread-safe and should be populated before it is used.
    */
   operator fun invoke(block: ClassName.() -> Unit) = this.block()

   /** Registers name for specified class and all subclasses that do not have any name registered. */
   infix fun String.alias(type: KClass<*>) {
      names[type] = this
      cache.keys.removeIf { type.isSuperclassOf(it) }
   }

   /**
    * Registers name for exactly the specified class.
    *
    * Desirable mostly for  general classes, such as [java.lang.Object], [java.lang.Void] or [Nothing].
    */
   infix fun String.aliasExact(type: KClass<*>) {
      namesNoSuper[type] = this
      cache.keys.removeIf { type.isSuperclassOf(it) }
   }

   /**
    * Returns name of the class. Name computation is in order:
    *  *  registered name of class
    *  *  registered name of first superclass in inheritance order
    *  *  registered name of first interface (no order)
    *  *  human readable name derived from class
    *
    * Lazy and cached. O(1).
    */
   operator fun get(type: KClass<*>) = cache.computeIfAbsent(type) {
      null
         ?: namesNoSuper[it]
         ?: names.getElementOfSuperV(it)
         ?: of(it)
   }

   /** Returns [get] for the class of the specified instance or for [Nothing] class if instance is null. */
   fun getOf(instance: Any?): String = when (instance) {
      null -> get(Nothing::class)
      else -> get(instance::class)
   }

   companion object {

      private fun of(type: KClass<*>): String = when {
         type.java.isAnonymousClass -> of(type.java.superclass.asIs<Class<*>>().kotlin)  // anonymous classes have no human readable name, use superclass
         else -> type.simpleName ?: type.jvmName
      }

   }

}